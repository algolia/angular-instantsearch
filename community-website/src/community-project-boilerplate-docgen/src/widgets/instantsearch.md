---
title: Widgets - Instantsearch
layout: widget.pug
---

# Instantsearch

## Description

Is the root component of all Angular InstantSearch implementations.

## Options

| Attributes    | Type            | Description
| -             | -               | -
| `config`      | `ConfigOptions` | The config object to pass to the instantsearch instance

Three parameters are required to get you started with instantsearch.js:

* `appId`: your algolia application id
* `apiKey`: the search key associated with your application
* `indexName`: the main index that you will use for your new search UI

Those parameters can be found in your [Algolia dashboard](https://www.algolia.com/api-keys).

## Code example

```js
@Component({
  selector: 'ais-app',
  template: `
    <ais-instantsearch
      [config]="{
        apiKey: '6be0576ff61c053d5f9a3225e2a90f76',
        appId: 'latency',
        indexName: 'instant_search'
      }"
    >
    </ais-instantsearch>
  `
})
export class AppComponent {}
```

## Config object options

`appId: string`
> The Algolia application ID

`apiKey: string`
> The Algolia search-only API key

`indexName: string`
> The name of the main index

`numberLocale?: string`
> The locale used to display numbers. This will be passed to `Number.prototype.toLocaleString()`

`searchFunction?: Function`
> A hook that will be called each time a search needs to be done, with the helper as a parameter. It’s your responsibility to call `helper.search()`. This option allows you to avoid doing searches at page load for example.

`createAlgoliaClient?: (algoliasearch: Function, appId: string, apiKey: string) => CustomClient`
> Allows you to provide your own algolia client instead of the one instantiated internally.
> Useful in situations where you need to setup complex mechanism on the client or if you need to share it easily.
> We forward `algoliasearch` which is the original algoliasearch module imported inside angular-instantsearch.

`searchParameters?: {}`
> Additional parameters to pass to the Algolia API.

`urlSync?: boolean | URLSyncOptions`
> Url synchronization configuration. Setting to true will synchronize the needed search parameters with the browser url.

#### URLSyncOptions

`mapping?: {}`
> Object used to define replacement query parameter to use in place of another. Keys are current query parameters and value the new value, e.g. `{ q: 'query' }`.

`threshold?: number`
> Idle time in ms after which a new state is created in the browser history. The default value is 700. The url is always updated at each keystroke but we only create a “previous search state” (activated when click on back button) every 700ms of idle time.

`trackedParameters?: string[]`
> Parameters that will be synchronized in the URL.
> Default value is `['query', 'attribute:\*', 'index', 'page', 'hitsPerPage']`. attribute:* means all the faceting attributes will be tracked. You can track only some of them by using `[…, 'attribute:color', 'attribute:categories']`. All other possible values are all the attributes of the Helper SearchParameters.

`useHash?: boolean`
> If set to true, the url will be hash based. Otherwise, it’ll use the query parameters using the modern history API.

`getHistoryState?: Function`
> Pass this function to override the default history API state we set to null. For example this could be used to force passing `{turbolinks: true}` to the history API every time we update it.

#### Events

You can register to the `(change)` event on the `<ais-instantsearch>` which will be fired after every new search. This event is useful for instance to push data to Google Analytics to know more about your users search:

```html
<ais-instantsearch
  [config]="{...}"
  (change)="onSearchChange($event)"
>
</ais-instantsearch>
```

```js
onSearchChange({ results, state }: { results: {}, state: {} }) {
  // Do what ever you need with the results or the state of search
}
```
