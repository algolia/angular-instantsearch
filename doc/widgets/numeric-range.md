## `<ng-ais-numeric-range />`

Numeric range widget allows a user to select a numeric range using a minimum and maximum input.

| Attributes       | Type       | Description
| -                | -          | -
| `attributeName`  | `string`   | Name of the attribute for faceting
| `min?`           | `number`   | Minimal value, default to automatically computed
| `max?`           | `number`   | Maximal value, default to automatically computed
| `precision?`     | `number`   | Number of digits after decimal point to use
| `header?`        | `string`   | Displays text before widget
| `footer?`        | `string`   | Displays text after widget
