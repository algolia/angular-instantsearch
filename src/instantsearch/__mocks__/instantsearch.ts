import { Component } from "@angular/core";

@Component({
  selector: "ais-instantsearch",
  template: "<ng-content></ng-content>"
})
export class NgAisInstantSearch {}
