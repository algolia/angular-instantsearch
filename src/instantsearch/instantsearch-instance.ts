import { Injectable, Inject, PLATFORM_ID } from "@angular/core";
import { isPlatformBrowser } from "@angular/common";
import { HttpClient, HttpHeaders } from "@angular/common/http";

import instantsearch from "instantsearch.js/es";
import * as algoliasearchProxy from "algoliasearch/index";
import * as encodeProxy from "querystring-es3/encode";

import { each } from "lodash-es";

import { Widget } from "../base-widget";

// AOT + Rollup workaround
// https://github.com/rollup/rollup/issues/1267#issuecomment-296395734

const algoliasearch = algoliasearchProxy.default || algoliasearchProxy;
const encode = encodeProxy.default || encodeProxy;

export type InstantSearchConfig = {
  appId: string;
  apiKey: string;
  indexName: string;

  numberLocale?: string;
  searchFunction?: () => void;
  createAlgoliaClient?: (
    algoliasearch: Function,
    appId: string,
    apiKey: string
  ) => object;
  searchParameters?: object | void;
  urlSync?:
    | boolean
    | {
        mapping?: object;
        threshold?: number;
        trackedParameters?: string[];
        useHash?: boolean;
        getHistoryState?: () => object;
      };
};

export class InstantSearchInstance {
  public start: () => void;

  public addWidget: (widget: Widget) => void;
  public addWidgets: (widgets: Widget[]) => void;

  public removeWidget: (widget: Widget) => void;
  public removeWidgets: (widgets: Widget[]) => void;

  // EventEmmiter
  public on: (eventName: string, callback: Function) => void;
  public removeListener: (eventName: string, callback: Function) => void;

  public helper: {
    lastResults: Object;
    state: Object;
  };
}

@Injectable()
export class NgAisInstance {
  private instance?: InstantSearchInstance;

  constructor(
    @Inject(PLATFORM_ID) private platformId: Object,
    private http: HttpClient
  ) {}

  public init(config: InstantSearchConfig) {
    // add default searchParameters with highlighting config
    if (!config.searchParameters) config.searchParameters = {};
    Object.assign(config.searchParameters, {
      highlightPreTag: "__ais-highlight__",
      highlightPostTag: "__/ais-highlight__"
    });

    const createAlgoliaClient = (_, appId, apiKey) => {
      const client = algoliasearch(appId, apiKey, {});

      client._request = (rawUrl, opts) => {
        let headers = new HttpHeaders();

        headers = headers.set(
          "content-type",
          opts.method === "POST"
            ? "application/x-www-form-urlencoded"
            : "application/json"
        );

        headers = headers.set("accept", "application/json");

        const url =
          rawUrl + (rawUrl.includes("?") ? "&" : "?") + encode(opts.headers);

        return new Promise((resolve, reject) => {
          this.http
            .request(opts.method, url, {
              headers,
              body: opts.body,
              observe: "response"
            })
            .subscribe(
              resp =>
                resolve({
                  statusCode: resp.status,
                  body: resp.body,
                  headers: resp.headers
                }),
              resp =>
                reject({
                  statusCode: resp.status,
                  body: resp.body,
                  headers: resp.headers
                })
            );
        });
      };

      return client;
    };

    Object.assign(config, { createAlgoliaClient });

    // remove URLSync widget if on SSR
    if (!isPlatformBrowser(this.platformId)) {
      config.urlSync = false;
    }

    this.instance = instantsearch(config);
  }

  public start() {
    this.instance.start();
  }

  public addWidget(widget: Widget) {
    this.instance.addWidget(widget);
  }

  public addWidgets(widgets: Widget[]) {
    this.instance.addWidgets(widgets);
  }

  public removeWidget(widget: Widget) {
    this.instance.removeWidget(widget);
  }

  public removeWidgets(widgets: Widget[]) {
    this.instance.removeWidgets(widgets);
  }

  public on(eventName: string, callback: Function) {
    this.instance.on(eventName, callback);
  }

  public off(eventName: string, callback: Function) {
    this.instance.removeListener(eventName, callback);
  }

  getResults() {
    return this.instance.helper.lastResults;
  }

  getState() {
    return this.instance.helper.state;
  }
}
