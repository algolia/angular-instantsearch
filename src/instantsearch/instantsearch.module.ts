import { CommonModule } from "@angular/common";
import { ModuleWithProviders, NgModule } from "@angular/core";
import { HttpClientModule } from "@angular/common/http";

import { NgAisInstantSearch } from "./instantsearch";

@NgModule({
  declarations: [NgAisInstantSearch],
  entryComponents: [NgAisInstantSearch],
  exports: [NgAisInstantSearch],
  imports: [CommonModule, HttpClientModule]
})
export class NgAisInstantSearchModule {
  public static forRoot(): ModuleWithProviders {
    return {
      ngModule: NgAisInstantSearchModule,
      providers: []
    };
  }
}
