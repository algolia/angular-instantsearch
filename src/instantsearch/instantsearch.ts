import {
  AfterViewInit,
  Component,
  Input,
  OnInit,
  OnDestroy,
  Output,
  EventEmitter,
  Inject,
  PLATFORM_ID
} from "@angular/core";
import { isPlatformBrowser } from "@angular/common";

import * as algoliasearchProxy from 'algoliasearch'
import instantsearch from "instantsearch.js/es";

import { Widget } from "../base-widget";
import { VERSION } from "../version";

const algoliasearch = algoliasearchProxy.default || algoliasearchProxy

export type InstantSearchConfig = {
  appId?: string;
  apiKey?: string;
  indexName: string;

  numberLocale?: string;
  searchFunction?: () => void;
  createAlgoliaClient?: (
    algoliasearch: Function,
    appId: string,
    apiKey: string
  ) => object;
  searchClient?: object;
  searchParameters?: object | void;
  urlSync?:
    | boolean
    | {
        mapping?: object;
        threshold?: number;
        trackedParameters?: string[];
        useHash?: boolean;
        getHistoryState?: () => object;
      };
};

export class InstantSearchInstance {
  public start: () => void;

  public addWidget: (widget: Widget) => void;
  public addWidgets: (widgets: Widget[]) => void;

  public removeWidget: (widget: Widget) => void;
  public removeWidgets: (widgets: Widget[]) => void;

  // EventEmmiter
  public on: (eventName: string, callback: Function) => void;
  public removeListener: (eventName: string, callback: Function) => void;

  public helper: {
    lastResults: Object;
    state: Object;
  };

  public dispose: () => void;
}

@Component({
  selector: "ais-instantsearch",
  template: `<ng-content></ng-content>`
})
export class NgAisInstantSearch implements AfterViewInit, OnInit, OnDestroy {
  @Input() public config: InstantSearchConfig;
  @Input() public instanceName: string = "default";

  @Output()
  change: EventEmitter<{ results: {}; state: {} }> = new EventEmitter<{
    results: {};
    state: {};
  }>();

  public instantSearchInstance: InstantSearchInstance;

  constructor(@Inject(PLATFORM_ID) private platformId: Object) {}

  public ngOnInit() {
    this.createInstantSearchInstance(this.config);
  }

  public ngAfterViewInit() {
    this.instantSearchInstance.start();
  }

  public ngOnDestroy() {
    this.instantSearchInstance.removeListener("render", this.onRender);
    this.instantSearchInstance.dispose();
  }

  public createInstantSearchInstance(config: InstantSearchConfig) {
    // add default searchParameters with highlighting config
    if (!config.searchParameters) config.searchParameters = {};
    Object.assign(config.searchParameters, {
      highlightPreTag: "__ais-highlight__",
      highlightPostTag: "__/ais-highlight__"
    });

    // remove URLSync widget if on SSR
    if (!isPlatformBrowser(this.platformId)) {
      config.urlSync = false;
    }

    // custom algolia client agent
    if (!config.searchClient && !config.createAlgoliaClient) {
      const client = algoliasearch(config.appId, config.apiKey);
      client.addAlgoliaAgent(`angular-instantsearch ${VERSION}`);

      config.searchClient = client;
      config.appId = undefined;
      config.apiKey = undefined;
    }

    this.instantSearchInstance = instantsearch(config);
    this.instantSearchInstance.on("render", this.onRender);
  }

  public addWidget(widget: Widget) {
    this.instantSearchInstance.addWidget(widget);
  }

  public removeWidget(widget: Widget) {
    this.instantSearchInstance.removeWidget(widget);
  }

  onRender = () => {
    this.change.emit({
      results: this.instantSearchInstance.helper.lastResults,
      state: this.instantSearchInstance.helper.state
    });
  };
}
