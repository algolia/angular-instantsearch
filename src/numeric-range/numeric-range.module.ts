import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgAisNumericRange } from "./numeric-range";

@NgModule({
  declarations: [NgAisNumericRange],
  entryComponents: [NgAisNumericRange],
  exports: [NgAisNumericRange],
  imports: [CommonModule]
})
export class NgAisNumericRangeModule {}
