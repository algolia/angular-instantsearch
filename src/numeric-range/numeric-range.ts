import { Component, Input } from "@angular/core";
import { connectRange } from "instantsearch.js/es/connectors";
import { noop } from "lodash-es";

import { BaseWidget } from "../base-widget";
import { NgAisInstance } from "../instantsearch/instantsearch-instance";
import { parseNumberInput } from "../utils";

export type NumericRangeState = {
  range: { min?: number; max?: number };
  refine: Function;
  start: number[];
};

@Component({
  selector: "ng-ais-numeric-range",
  template: `
    <div [class]="cx()">
      <div [class]="cx('body')">
        <form
          [class]="cx('form')"
          (submit)="handleSubmit($event)"
          novalidate
        >
          <label [class]="cx('label')">
            <span [class]="cx('currency')">{{currency}} </span>
            <input
              [class]="cx('input')"
              type="number"
              [min]="state.range.min"
              [max]="state.range.max"
              [placeholder]="state.range.min"
              [value]="minInputValue"
              [step]="step"
              (change)="handleChange($event, 'min')"
            />
          </label>

          <span [class]="cx('separator')"> {{separator}} </span>

          <label [class]="cx('label')">
            <span [class]="cx('currency')">{{currency}} </span>
            <input
              [class]="cx('input')"
              type="number"
              [min]="state.range.min"
              [max]="state.range.max"
              [placeholder]="state.range.max"
              [value]="maxInputValue"
              [step]="step"
              (change)="handleChange($event, 'max')"
            />
          </label>

          <button
            [class]="cx('submit')"
            (click)="handleSubmit($event)"
          >
            {{submitLabel}}
          </button>
        </form>
      </div>
    </div>
  `
})
export class NgAisNumericRange extends BaseWidget {
  // render options
  @Input() public currency: string = "$";
  @Input() public separator: string = "to";
  @Input() public submitLabel: string = "Go";

  // connector options
  @Input() public attributeName: string;
  @Input() public min?: number | string;
  @Input() public max?: number | string;
  @Input() public precision?: number | string = 2;

  // inner state
  public minInputValue: number | string = "";
  public maxInputValue: number | string = "";

  get step() {
    return 1 / Math.pow(10, parseNumberInput(this.precision));
  }

  public state: NumericRangeState = {
    range: { min: undefined, max: undefined },
    refine: noop,
    start: [0, 0]
  };

  constructor(searchInstance: NgAisInstance) {
    super(searchInstance, "RangeSlider");
  }

  public ngOnInit() {
    this.createWidget(connectRange, {
      attributeName: this.attributeName,
      max: parseNumberInput(this.max),
      min: parseNumberInput(this.min),
      precision: parseNumberInput(this.precision)
    });

    super.ngOnInit();
  }

  public handleChange(event: any, type: string) {
    const value = parseNumberInput(event.target.value);

    if (type === "min") {
      this.minInputValue = value;
    } else {
      this.maxInputValue = value;
    }
  }

  public handleSubmit(event: MouseEvent | KeyboardEvent) {
    event.preventDefault();
    this.state.refine([this.minInputValue, this.maxInputValue]);
  }
}
