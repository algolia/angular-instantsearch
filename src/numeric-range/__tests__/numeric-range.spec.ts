import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../../instantsearch/instantsearch.module";
import { NgAisNumericRangeModule } from "../numeric-range.module";
import { NgAisNumericRange } from "../numeric-range";

jest.mock("../../base-widget");

const defaultState = {
  range: { min: 0, max: 100 },
  start: [0, 100],
  refine: jest.fn()
};

const render = (state?: {}) => {
  const fixture = TestBed.createComponent(NgAisNumericRange);

  if (state) {
    fixture.componentInstance.updateState({ ...defaultState, ...state });
  }

  fixture.detectChanges();
  return fixture;
};

describe("NumericRange", () => {
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [],
      imports: [NgAisInstantSearchModule.forRoot(), NgAisNumericRangeModule]
    });
  });

  it("renders markup without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("renders markup with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it("should update `min/max InputValue`", () => {
    const fixture = render({});

    const [
      minInput,
      maxInput
    ] = fixture.debugElement.nativeElement.querySelectorAll("input");

    minInput.value = "20";
    minInput.dispatchEvent(new Event("change"));

    maxInput.value = "50";
    maxInput.dispatchEvent(new Event("change"));

    expect(fixture.componentInstance.minInputValue).toBe(20);
    expect(fixture.componentInstance.maxInputValue).toBe(50);
  });

  it("should call renfine when submitting form", () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const [
      minInput,
      maxInput
    ] = fixture.debugElement.nativeElement.querySelectorAll("input");

    minInput.value = "20";
    minInput.dispatchEvent(new Event("change"));

    maxInput.value = "50";
    maxInput.dispatchEvent(new Event("change"));

    expect(fixture.componentInstance.minInputValue).toBe(20);
    expect(fixture.componentInstance.maxInputValue).toBe(50);

    const submitBtn = fixture.debugElement.nativeElement.querySelector(
      "button"
    );
    submitBtn.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith([20, 50]);
  });
});
