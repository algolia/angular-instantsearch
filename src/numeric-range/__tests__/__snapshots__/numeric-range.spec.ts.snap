// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericRange renders markup with state 1`] = `
<ng-ais-numeric-range
  currency={[Function String]}
  cx={[Function Function]}
  maxInputValue=""
  minInputValue=""
  precision={[Function Number]}
  searchInstance={[Function NgAisInstance]}
  separator={[Function String]}
  state={[Function Object]}
  submitLabel={[Function String]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-RangeSlider"
  >
    
        
    <div
      class="ais-RangeSlider-body ais-body"
    >
      
          
      <form
        class="ais-RangeSlider-form"
        novalidate=""
      >
        
            
        <label
          class="ais-RangeSlider-label"
        >
          
              
          <span
            class="ais-RangeSlider-currency"
          >
            $ 
          </span>
          
              
          <input
            class="ais-RangeSlider-input"
            max="100"
            min="0"
            placeholder="0"
            step="0.01"
            type="number"
          />
          
            
        </label>
        
  
            
        <span
          class="ais-RangeSlider-separator"
        >
           to 
        </span>
        
  
            
        <label
          class="ais-RangeSlider-label"
        >
          
              
          <span
            class="ais-RangeSlider-currency"
          >
            $ 
          </span>
          
              
          <input
            class="ais-RangeSlider-input"
            max="100"
            min="0"
            placeholder="100"
            step="0.01"
            type="number"
          />
          
            
        </label>
        
  
            
        <button
          class="ais-RangeSlider-submit"
        >
          
              Go
            
        </button>
        
          
      </form>
      
        
    </div>
    
      
  </div>
    
</ng-ais-numeric-range>
`;

exports[`NumericRange renders markup without state 1`] = `
<ng-ais-numeric-range
  currency={[Function String]}
  cx={[Function Function]}
  maxInputValue=""
  minInputValue=""
  precision={[Function Number]}
  searchInstance={[Function NgAisInstance]}
  separator={[Function String]}
  state={[Function Object]}
  submitLabel={[Function String]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-RangeSlider"
  >
    
        
    <div
      class="ais-RangeSlider-body ais-body"
    >
      
          
      <form
        class="ais-RangeSlider-form"
        novalidate=""
      >
        
            
        <label
          class="ais-RangeSlider-label"
        >
          
              
          <span
            class="ais-RangeSlider-currency"
          >
            $ 
          </span>
          
              
          <input
            class="ais-RangeSlider-input"
            max="undefined"
            min="undefined"
            placeholder="undefined"
            step="0.01"
            type="number"
          />
          
            
        </label>
        
  
            
        <span
          class="ais-RangeSlider-separator"
        >
           to 
        </span>
        
  
            
        <label
          class="ais-RangeSlider-label"
        >
          
              
          <span
            class="ais-RangeSlider-currency"
          >
            $ 
          </span>
          
              
          <input
            class="ais-RangeSlider-input"
            max="undefined"
            min="undefined"
            placeholder="undefined"
            step="0.01"
            type="number"
          />
          
            
        </label>
        
  
            
        <button
          class="ais-RangeSlider-submit"
        >
          
              Go
            
        </button>
        
          
      </form>
      
        
    </div>
    
      
  </div>
    
</ng-ais-numeric-range>
`;
